### PR TITLE
Add shared project (for data egress protection)

### PR DESCRIPTION
**Pull request by illusional**
Configurable on the dataset's config:

```yaml
  datasets:enable_egress_project: "true"
  datasets:egress_budget: 123
```

- Creates a new project with format `cpg-{dataset}-egress` to avoid any name clashes
- Creates a budget for that project
- Creates a service account in that new project
- Allows that service-account to access requester-pays buckets
- Allow that service-account to access the _release_ bucket on the parent project.

### CODE REVIEW

File: stack/__main__.py [hunk 1]
```diff
@@ -650,6 +657,117 @@ def add_access_group_cache_as_secret_member(secret, resource_prefix: str):
...
 
+    if enable_egress_project:
+        if not release_bucket:
+            raise ValueError(
+                'Requested egress project, but no bucket is available to egress from'
+            )
+
+        egress_buckets = {'release': release_bucket}
+
+        project_name = f'cpg-{dataset}-egress'
```
Comment (lgruen) [inline stack/__main__.py:668]: How about we make this a bit more generic, so we can potentially use this for compute as well in the future? Can we make the name consistent with the existing project names?

```suggestion
        project_name = f'{project_id}-shared'
```

File: stack/__main__.py [hunk 2]
```diff
@@ -650,6 +657,117 @@ def add_access_group_cache_as_secret_member(secret, resource_prefix: str):
             member=pulumi.Output.concat('group:', release_access_group.group_key.id),
         )
 
+    if enable_egress_project:
+        if not release_bucket:
+            raise ValueError(
+                'Requested egress project, but no bucket is available to egress from'
+            )
```
Comment (lgruen) [inline stack/__main__.py:664]: We could skip that, as the same logic will apply once we share `main` buckets (with Requester Pays enabled).

**Review by lgruen — CHANGES_REQUESTED** [review]

File: stack/__main__.py (hunk 2 above)
Comment (illusional) [reply stack/__main__.py:664]: Are you okay if I leave this here until that's implemented, already caught me once :)

File: stack/__main__.py (hunk 1 above)
Comment (lgruen) [reply stack/__main__.py]: Do you want to rename all the other "egress" occurrences as well?

**Review by lgruen — CHANGES_REQUESTED** [review]
Do you want to rename all the other "egress" occurrences as well?

File: stack/__main__.py (hunk 1 above)
Comment (illusional) [reply stack/__main__.py]: Ah right, I see sorry.

Comment (illusional) [reply stack/__main__.py]: Tested and pushed new changes :)

File: stack/__main__.py [hunk 3]
```diff
@@ -650,6 +657,117 @@ def add_access_group_cache_as_secret_member(secret, resource_prefix: str):
             member=pulumi.Output.concat('group:', release_access_group.group_key.id),
         )
 
+    if enable_shared_project:
+        if not release_bucket:
+            raise ValueError(
+                'Requested shared project, but no bucket is available to shared from'
```
Comment (lgruen) [inline stack/__main__.py:663]:
```suggestion
                'Requested shared project, but no bucket is available to share'
```

File: stack/__main__.py [hunk 4]
```diff
@@ -650,6 +657,117 @@ def add_access_group_cache_as_secret_member(secret, resource_prefix: str):
...
+        shared_service_account = gcp.serviceaccount.Account(
+            'budget-shared-service-account',
+            account_id=f'shared',
+            opts=pulumi.resource.ResourceOptions(
+                depends_on=[shared_cloudidentity],
+            ),
+            project=shared_project.name,
+        )
+
+        # Allow the usage of requester-pays buckets.
```
Comment (lgruen) [inline stack/__main__.py:751]:
```suggestion
        # Allow the usage of Requester Pays buckets.
```

**Review by lgruen — APPROVED** [review]

File: stack/__main__.py [hunk 5]
```diff
@@ -650,6 +657,117 @@ def add_access_group_cache_as_secret_member(secret, resource_prefix: str):
...
+        )
+        cloudbillingbudgets = gcp.projects.Service(
+            'cloudbillingbudgets-service',
+            service='billingbudgets.googleapis.com',
+            disable_on_destroy=False,
+            project=project_id,
+            opts=pulumi.resource.ResourceOptions(depends_on=[cloudbilling]),
+        )
+
+        # create project to cover shared
```
[the remark this answers is not in the document]
Comment (illusional) [reply stack/__main__.py]: Sorry, this was sloppy from me

**Comment by lgruen** [comment]
As this is the first time we're creating `release` buckets, should we use this opportunity to remove the `-requester-pays` suffix from the bucket name if we'll be enabling Requester Pays on all buckets anyhow?

**Comment by illusional** [comment]
Addressed requester-pays release bucket naming in another PR, merging this for now :)

**Pull request merged**